mllib/src/main/scala/org/apache/spark/ml/feature/Interaction.scala

```diff
@@ -0,0 +1,243 @@
+ * produced.
+ *
+ * For example, given the inputs Double(2), Vector(3, 4), the result would be Vector(6, 8) if
+ * all columns were numeric. If the first input was nominal with four different values, the result
+ * would then be Vector(0, 0, 0, 0, 3, 4, 0, 0).
+ *
+ * See https://stat.ethz.ch/R-manual/R-devel/library/base/html/formula.html for more
+ * information about interactions in R formulae.
+ */
+@Experimental
+class Interaction(override val uid: String) extends Transformer
+  with HasInputCols with HasOutputCol {
+
+  def this() = this(Identifiable.randomUID("interaction"))
+
+  /** @group setParam */
+  def setInputCols(values: Array[String]): this.type = set(inputCols, values)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  // optimistic schema; does not contain any ML attributes
+  override def transformSchema(schema: StructType): StructType = {
+    checkParams()
+    StructType(schema.fields :+ StructField($(outputCol), new VectorUDT, true))
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    checkParams()
+    val fieldIterators = getIterators(dataset)
+
+    def interactFunc = udf { row: Row =>
+      var indices = ArrayBuilder.make[Int]
+      var values = ArrayBuilder.make[Double]
+      var size = 1
+      indices += 0
+      values += 1.0
+      var fieldIndex = row.length - 1
+      while (fieldIndex >= 0) {
+        val prevIndices = indices.result()
+        val prevValues = values.result()
+        val prevSize = size
+        val currentIterator = fieldIterators(fieldIndex)
+        indices = ArrayBuilder.make[Int]
+        values = ArrayBuilder.make[Double]
+        size *= currentIterator.size
+        currentIterator.foreachActive(row(fieldIndex), (i, a) => {
+          var j = 0
+          while (j < prevIndices.length) {
+            indices += prevIndices(j) + i * prevSize
+            values += prevValues(j) * a
+            j += 1
+          }
+        })
+        fieldIndex -= 1
+      }
+      Vectors.sparse(size, indices.result(), values.result()).compressed
+    }
+
+    val args = $(inputCols).map { c =>
+      dataset.schema(c).dataType match {
+        case DoubleType => dataset(c)
+        case _: VectorUDT => dataset(c)
+        case _: NumericType | BooleanType => dataset(c).cast(DoubleType)
+      }
+    }
+    val attrs = generateAttrs($(inputCols).map(col => dataset.schema(col)))
+    dataset.select(
+      col("*"),
+      interactFunc(struct(args: _*)).as($(outputCol), attrs.toMetadata()))
+  }
+
+  private def getIterators(dataset: DataFrame): Array[EncodingIterator] = {
+    def getCardinality(attr: Attribute): Int = {
+      attr match {
+        case nominal: NominalAttribute =>
+          nominal.getNumValues.getOrElse(
+            throw new SparkException("Nominal fields must have attr numValues defined."))
+        case _ =>
+          0  // treated as numeric
```

Comment:
This is used to calculate the output index of iteration features. Maybe we can change `cardinality` to `numFeatures`. Then the output is `numValues` for a nominal feature, 1 for a binary feature, and 1 for a numeric feature.

Btw, (not in this PR) it might be useful to store only the nonzeros because most features would be numeric features. If there are millions of features, this cardinality array becomes very large.